tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -1877,40 +1845,30 @@ void CGMSHLSLRuntime::AddHLSLFunctionInfo(Function *F, const FunctionDecl *FD) {
           FD->getAttr<HLSLNodeLocalRootArgumentsTableIndexAttr>()) {
     if (isNode)
       funcProps->Node.LocalRootArgumentsTableIndex = pAttr->getIndex();
-    else
-      ReportMissingNodeDiag(Diags, pAttr);
   }
   if (const auto *pAttr = FD->getAttr<HLSLNodeShareInputOfAttr>()) {
     if (isNode) {
       funcProps->NodeShaderSharedInput.Name = pAttr->getName().str();
       funcProps->NodeShaderSharedInput.Index = pAttr->getArrayIndex();
-    } else {
-      ReportMissingNodeDiag(Diags, pAttr);
     }
   }
   if (const auto *pAttr = FD->getAttr<HLSLNodeDispatchGridAttr>()) {
     if (isNode) {
       funcProps->Node.DispatchGrid[0] = pAttr->getX();
       funcProps->Node.DispatchGrid[1] = pAttr->getY();
       funcProps->Node.DispatchGrid[2] = pAttr->getZ();
-    } else {
-      ReportMissingNodeDiag(Diags, pAttr);
     }
   }
   if (const auto *pAttr = FD->getAttr<HLSLNodeMaxDispatchGridAttr>()) {
     if (isNode) {
       funcProps->Node.MaxDispatchGrid[0] = pAttr->getX();
       funcProps->Node.MaxDispatchGrid[1] = pAttr->getY();
       funcProps->Node.MaxDispatchGrid[2] = pAttr->getZ();
-    } else {
-      ReportMissingNodeDiag(Diags, pAttr);
     }
   }
   if (const auto *pAttr = FD->getAttr<HLSLNodeMaxRecursionDepthAttr>()) {
     if (isNode)
       funcProps->Node.MaxRecursionDepth = pAttr->getCount();
-    else
-      ReportMissingNodeDiag(Diags, pAttr);
   }
   if (!FD->getAttr<HLSLNumThreadsAttr>()) {
     if (isNode) {
```

Comment:
GitHub won't let me write a suggestion for this because this block spans across a lot of code, but there's a really gross pattern here that we should clean up.

We have a chain of statements here that are basically:

```c++
if (/*getAttr succeedes*/)
  if (isNode)
    // do something
```

Clang's getAttr isn't crazy slow, but it is silly to do a bunch of attribute lookups if we're never going to do anything with the result.

Can we instead rewrite this to remove all the inner `if (isNode)` checks and just fold all the code up under the `if (isNode)` block on line 1816 above?

Reply:
Done